plutus-metatheory/src/Cost/Size.lagda.md

```diff
@@ -0,0 +1,114 @@
+# Size of Constants
+
+```
+module Cost.Size where 
+
+```
+
+# Imports
+
+```
+open import Data.Nat using (ℕ;zero;suc;_+_;_*_;_∸_)
+open import Data.Nat.DivMod using (_/_)
+open import Data.Integer using (ℤ;∣_∣)
+open import Data.String using (length)
+
+open import Utils using (_×_;_,_;_∷_;[];DATA;List;ByteString)
+open DATA
+
+open import Builtin using (lengthBS)
+open import Builtin.Signature using (_⊢♯)
+open _⊢♯
+
+open import Builtin.Constant.AtomicType using (AtomicTyCon)
+open AtomicTyCon
+open import Cost.Base 
+open import Untyped.CEK using (Value;V-con)
+open import RawU using (TmCon;tmCon) 
+
+open import Data.Vec using (Vec;[];_∷_) 
+                     renaming (lookup to lookupVec)
+
+```
+
+## Memory usage of type constants
+
+For each type constant we calculate its size, as a measure of memory usage.
+
+First we bring some functions from Haskell world.
+
+```
+postulate ℕtoWords : ℤ → CostingNat 
+postulate g1ElementCost : CostingNat
+postulate g2ElementCost : CostingNat
+postulate mlResultElementCost : CostingNat 
+
+{-# FOREIGN GHC {-# LANGUAGE MagicHash #-} #-}
+{-# FOREIGN GHC import GHC.Exts (Int (I#)) #-}
+{-# FOREIGN GHC import GHC.Integer  #-}
+{-# FOREIGN GHC import GHC.Integer.Logarithms  #-}
+{-# FOREIGN GHC import GHC.Prim #-}
+{-# FOREIGN GHC import PlutusCore.Crypto.BLS12_381.G1 as BLS12_381.G1 #-}
+{-# FOREIGN GHC import PlutusCore.Crypto.BLS12_381.G2 as BLS12_381.G2 #-}
+{-# FOREIGN GHC import PlutusCore.Crypto.BLS12_381.Pairing as BLS12_381.Pairing #-}
+
+{-# COMPILE GHC ℕtoWords = \i -> fromIntegral $ I# (integerLog2# (abs i) `quotInt#` integerToInt 64) + 1 #-}
+{-# COMPILE GHC g1ElementCost = toInteger (BLS12_381.G1.memSizeBytes `div` 8) #-}
+{-# COMPILE GHC g2ElementCost = toInteger (BLS12_381.G2.memSizeBytes `div` 8) #-}
+{-# COMPILE GHC mlResultElementCost = toInteger (BLS12_381.Pairing.mlResultMemSizeBytes `div` 8) #-}
+```
+
+For each constant we return the corresponding size.
+
+```
+byteStringSize : ByteString → CostingNat 
+byteStringSize x = let n = ∣ lengthBS x ∣ in ((n ∸ 1) / 8) + 1
+
+-- cost of a Data node
+dataNodeMem : CostingNat 
+dataNodeMem = 4
+
+sizeData : DATA → CostingNat 
+sizeDataList : List DATA → CostingNat 
+sizeDataDataList : List (DATA × DATA) → CostingNat
+
+sizeData (ConstrDATA _ xs) = dataNodeMem + sizeDataList xs
+sizeData (MapDATA []) = dataNodeMem
+sizeData (MapDATA xs) = dataNodeMem + sizeDataDataList xs
+sizeData (ListDATA xs) = dataNodeMem + sizeDataList xs
+sizeData (iDATA x) = dataNodeMem +  ℕtoWords x
+sizeData (bDATA x) = dataNodeMem + byteStringSize x 
+
+-- The following two functions are stupid but
+-- they make the termination checker happy
+sizeDataDataList [] = 0
+sizeDataDataList ((x , y) ∷ xs) = sizeData x + sizeData y + sizeDataDataList xs
+
+sizeDataList [] = 0
+sizeDataList (x ∷ xs) = sizeData x + sizeDataList xs
+
+defaultConstantMeasure : TmCon → CostingNat
+defaultConstantMeasure (tmCon (atomic aInteger) x) = ℕtoWords x
```

Comment:
The fact that these measures have to be the same as the Haskell versions is maybe a source of fragility since we might forget to keep them synchronised (although I suppose that budget tests might help with that).  Maybe it's unavoidable though.  If we ever want to prove something about costs we may need to be able to see exactly what the costing functions look like (so we can check that they're monotone or always positive or something),  and if we were somehow  accessing the Haskell versions directly instead then that might not be possible (I'm just thinking aloud here).

Reply:
I think it's ok to abstract over constants. We probably don't need to know the precise value to reason over them (unless we want to calculate formally that something is an optimization).